packages/core/src/fields/resolve-hooks.ts

```diff
@@ -0,0 +1,69 @@
+import {
+  type BaseListTypeInfo,
+  type FieldHooks,
+  type MaybePromise
+} from '../types'
+
+// force new syntax for built-in fields
+//   and block hooks from using resolveInput, they should use GraphQL resolvers
+export type InternalFieldHooks<ListTypeInfo extends BaseListTypeInfo> =
+  Omit<FieldHooks<ListTypeInfo>, 'validateInput' | 'validateDelete' | 'resolveInput'>
+
+/** @deprecated, TODO: remove in breaking change */
+function resolveValidateHooks <ListTypeInfo extends BaseListTypeInfo> ({
+  validate,
+  validateInput,
+  validateDelete
+}: FieldHooks<ListTypeInfo>): Exclude<FieldHooks<ListTypeInfo>["validate"], Function> | undefined {
+  if (validateInput || validateDelete) {
+    return {
+      create: validateInput,
+      update: validateInput,
+      delete: validateDelete,
+    }
+  }
+
+  if (!validate) return
+  if (typeof validate === 'function') {
+    return {
+      create: validate,
+      update: validate,
+      delete: validate
+    }
+  }
+
+  return validate
+}
+
+function merge <
+  R,
+  A extends (r: R) => MaybePromise<void>,
+  B extends (r: R) => MaybePromise<void>
+> (a?: A, b?: B) {
+  if (!a && !b) return undefined
+  return async (args: R) => {
+    await a?.(args)
+    await b?.(args)
+  }
+}
+
+export function mergeFieldHooks <ListTypeInfo extends BaseListTypeInfo> (
+  builtin?: InternalFieldHooks<ListTypeInfo>,
+  hooks?: FieldHooks<ListTypeInfo>,
+) {
+  if (hooks === undefined) return builtin
+  if (builtin === undefined) return hooks
+
+  const builtinValidate = resolveValidateHooks(builtin)
+  const hooksValidate = resolveValidateHooks(hooks)
+  return {
+    ...hooks,
+    beforeOperation: merge(builtin.beforeOperation, hooks.beforeOperation),
```

Comment:
@acburdine I did notice you had `hooks.beforeOperation` before the built-ins, was that intentional?

Reply:
Resolved, this was intentional, and I have added a comment as such